### PR TITLE
Add MySQL support to backfill script

### PR DESCRIPTION
**Pull request by cmurphy**
Rename backfill-redis to backfill-index and add support for when MySQL
is used as the index storage backend.

Some Redis-specific parameters are renamed to clearly differentiate them
from MySQL parameters. MySQL connection parameters mirror the
rekor-server paramters, using a single --dsn flag instead of separate
host, port, password, etc flags.

To do:

- [x] Rename backfill-redis and combine both backfill scripts into one?
- [x] Functional tests for backfill-redis
- [x] Functional tests for backfill-mysql

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -118,6 +118,37 @@ jobs:
...
+    runs-on: ubuntu-latest
+    needs: build
+
+    steps:
+      - uses: actions/checkout@1d96c772d19495a3b5c517cd2bc0cb401ea0529f # v4.1.3
+      - name: Extract version of Go to use
+        run: echo "GOVERSION=$(cat Dockerfile|grep golang | awk ' { print $2 } ' | cut -d '@' -f 1 | cut -d ':' -f 2 | uniq)" >> $GITHUB_ENV
+      - uses: actions/setup-go@0c52d547c9bc32b1aa3301fd7a9cb496313a4491 # v5.0.0
+        with:
+          go-version: ${{ env.GOVERSION }}
```
Comment (Hayden-IO) [inline .github/workflows/main.yml:131]: Can we use `go-version-file: go.mod` instead? Dockerfile and go.mod should stay in sync.

File: cmd/backfill-index/main.go [hunk 2]
```diff
@@ -87,11 +116,23 @@ func main() {
 		os.Exit(0)
 	}
 
-	if *redisHostname == "" {
-		log.Fatal("address must be set")
+	provider := providerUnset
```
Comment (Hayden-IO) [inline cmd/backfill-index/main.go:119]: nit, should we check that both mysqlDSN and redisHostname aren't set?

File: docker-compose.backfill-test.yml [hunk 3]
```diff
@@ -0,0 +1,61 @@
+#
+# Copyright 2022 The Sigstore Authors.
```
Comment (Hayden-IO) [inline docker-compose.backfill-test.yml]: 2024

**Review by Hayden-IO — COMMENTED** [review]
LGTM, fantastic work on the very thorough tests!

File: .github/workflows/main.yml (hunk 1 above)
Comment (cmurphy) [reply .github/workflows/main.yml:131]: I can, but that would make this job inconsistent with every other workflow in rekor, fulcio, and cosign, for example line 89 above. Do you want this changed everywhere?

Comment (Hayden-IO) [reply .github/workflows/main.yml:131]: Let's keep it consistent, we can change it in a follow up.

**Review by Hayden-IO — APPROVED** [review]

**Pull request merged**